File: content-processor/app/schemas/Metadata.py
from enum import Enum
from typing import Generic, Optional, TypeVar

from pydantic import BaseModel

# Specific metadata models for different content types


class YouTubeSpecificMd(BaseModel):
    video_id: str
    channel_name: str
    author_name: str
    chunk_id: str
    start_time: Optional[str] = None
    end_time: Optional[str] = None


class GitSpecificMd(BaseModel):
    repo_name: str
    repo_creator_name: str
    file_name: str
    programming_language: str
    chunk_type: str
    chunk_id: str


class MediaSpecificMd(BaseModel):
    type: str
    chunk_id: str
    start_time: Optional[float] = None
    end_time: Optional[float] = None


class ImageSpecificMd(BaseModel):
    width: int
    height: int
    format: str
    chunk_id: str


class TextSpecificMd(BaseModel):
    chunk_id: str
    url: str


class NoteSpecificMd(BaseModel):
    chunk_id: str


class NotionSpecificMd(BaseModel):
    chunk_id: str
    page_id: str


class MindMapSpecificMd(BaseModel):
    memory_count: int
    central_topic: str
    subtopics: list[str]
    chunk_id: str


class GDriveFileType(Enum):
    GDOC = "application/vnd.google-apps.document"
    GSHEET = "application/vnd.google-apps.spreadsheet"
    GSLIDE = "application/vnd.google-apps.presentation"
    PDF = "application/pdf"
    IMAGE = "image"
    AUDIO = "audio"
    VIDEO = "video"
    UNKNOWN = "unknown"


class GDriveSpecificMd(BaseModel):
    chunk_id: str
    file_id: str
    page_number: Optional[int]  # For slides
    sheet_name: Optional[str]   # For sheets


T = TypeVar('T')

# Main Metadata model that includes common fields and specific metadata


class Metadata(BaseModel, Generic[T]):
    user_id: str
    memId: str
    title: str
    description: str
    created_at: str
    last_updated: str
    tags: list[str]
    source: str
    language: str = "english"
    type: str
    content_hash: Optional[str] = None
    specific_desc: T
    ai_summary: Optional[str] = None
    ai_insights: Optional[str] = None
    related_memories: Optional[list[str]] = None

File: consumer/src/types/metadata.go
package types

import (
	"github.com/golang-jwt/jwt/v5"
)

// Metadata represents the main metadata model that includes common fields and specific metadata
type Metadata struct {
	UserID          string      `json:"user_id"`
	MemID           string      `json:"mem_id"`
	Title           string      `json:"title"`
	Description     string      `json:"description"`
	CreatedAt       string      `json:"created_at"`
	LastUpdated     string      `json:"last_updated"`
	Tags            []string    `json:"tags"`
	Source          string      `json:"source"`
	Language        string      `json:"language"`
	Type            string      `json:"type"`
	ContentHash     *string     `json:"content_hash,omitempty"`
	SpecificDesc    interface{} `json:"specific_desc"`
	AISummary       *string     `json:"ai_summary,omitempty"`
	AIInsights      *string     `json:"ai_insights,omitempty"`
	RelatedMemories []string    `json:"related_memories,omitempty"`
}

type Claims struct {
	UserId string `json:"user_id"`
	APIKey string `json:"api_key"`
	jwt.RegisteredClaims
}

type Task struct {
	Type    string      `json:"type"`
	Data    interface{} `json:"data"`
	Retries int         `json:"retries"`
}

type AgentResponse struct {
	Transcript string      `json:"transcript"`
	Chunks     []string    `json:"chunks"`
	Metadata   interface{} `json:"metadata"`
	UserID     string      `json:"userId"`
	MemoryID   string      `json:"memoryId"`
	Title      string      `json:"title"`
}

type AgentError struct {
	Error string `json:"error"`
}

type AgentResponseWrapper struct {
	Response *AgentResponse `json:"response"`
	Error    *AgentError    `json:"error"`
}

type User struct {
	ID    string
	Name  string
	Email string
}
